sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsRowLevelOperations.java

```diff
@@ -39,4 +45,22 @@ public interface SupportsRowLevelOperations extends Table {
    * @return the row-level operation builder
    */
   RowLevelOperationBuilder newRowLevelOperationBuilder(RowLevelOperationInfo info);
+
+
+  /**
+   * Calculate target table schema for MERGE INTO schema evolution.
+   * @param sourceTableSchema schema of the source table for MERGE INTO operation.
+   * @return new schema for the target table
+   */
+  default StructType mergeSchema(StructType sourceTableSchema) {
```

Comment:
There's two api's because Spark needs to first get the schema , for the resolution.  Then it executes later in extended-resolution.  Another option is to execute it earlier in the first pass, but not sure if there are concerns in doing it so early before analysis has a chance to succeed.